### PR TITLE
[gzip]: allow gzip to work w/ http2 backend w/o content-length

### LINKED ISSUE #14121: compression doesn't work with a streaming HTTP/2 upstream
*Title*: compression doesn't work with an HTTP/2 backend

*Description*:
If I understand existing code correctly then compression won't be applied to streaming responses returned by an HTTP/2-capable backend:
https://github.com/envoyproxy/envoy/blob/ca6e246336a6c454132b5c832d483b8e9aa10ca2/source/extensions/filters/http/common/compressor/compressor.cc#L396-L397

*Repro steps*:
Enable h2 to the backend, enable compressor, use streaming-capable upstream.

cc: @rojkov

### PR DESCRIPTION
**Pull request by tehnerd**
Commit Message:
this is an attempt to address issue in https://github.com/envoyproxy/envoy/issues/14121
tl;dr is that some http/2 backends does not send "content-length" header in replies, as http/2 spec do have the same info as a part of the frame (unfortunately it seems like there is no way to pass this value from the frame to the filter) and transfer-encoding=chunked (before this diff having that header/encoding was a prerequisite, if content-length is not defined) was removed from http/2 spec. 
As discussed in the issue itself, instead, if there is no content lengths header - we would try to gzip it by default.
this new behavior is controlled by runtime guarded feature `envoy.reloadable_features.enable_compression_without_chunked_header`

Additional Description:

Risk Level:
Low
Testing:
UTs + manual test w/ http2 backend
Docs Changes:
Release Notes:
Platform Specific Features:
[Optional Runtime guard:]
Fixes #14121 
[Optional Deprecated:]

### CODE REVIEW

File: source/extensions/filters/http/common/compressor/compressor.cc [hunk 1]
```diff
@@ -484,6 +490,14 @@ bool CompressorFilterConfig::DirectionConfig::isContentTypeAllowed(
   return true;
 }
 
+bool CompressorFilter::isResponseWithContent(Http::ResponseHeaderMap& headers) const {
+  const auto status = headers.Status();
+  if ((status != nullptr && status->value().getStringView() == "204") || head_request_) {
```
Comment (alyssawilk) [inline source/extensions/filters/http/common/compressor/compressor.cc:495]: Would this cause a behavioral change for proxied 101 responses?     Also I don't think 304 responses can have a body.  I wonder if it's worth refactoring the status code logic in the HTTP/1.1 codec to have a shared utility function.

Either way I think it's worth runtime guarding this feature as it affects the data plane.

**Review by alyssawilk — COMMENTED** [review]
Thanks for tackling this one.  A few thoughts to get you started:

File: source/extensions/filters/http/common/compressor/compressor.cc (hunk 1 above)
Comment (tehnerd) [reply source/extensions/filters/http/common/compressor/compressor.cc:502]: thanks for the review @alyssawilk . i've added new config option which enables this new behavior (i'm bad with naming though, so more than ready to change it if needed). as for "refactoring the status code logic" could you please elaborate on this? are you talking about some kind of utility like ``` hasContent(status_code) bool ``` which checks internally is it "204/304/something else or not"?

File: source/extensions/filters/http/common/compressor/compressor.cc [hunk 2]
```diff
@@ -109,6 +109,10 @@ CompressorFilterConfig::ResponseDirectionConfig::ResponseDirectionConfig(
           proto_config.has_response_direction_config()
               ? proto_config.response_direction_config().remove_accept_encoding_header()
               : proto_config.remove_accept_encoding_header()),
+      always_compress_content_if_exists_(
+          proto_config.has_response_direction_config()
```
Comment (htuch) [inline source/extensions/filters/http/common/compressor/compressor.cc:113]: This can be abbreviated to just `proto_config.response_direction_config().always_compress_content_if_exists()`, since proto has matching default semantics.

**Review by htuch — COMMENTED** [review]
/lgtm api

File: source/extensions/filters/http/common/compressor/compressor.cc (hunk 1 above)
Comment (rojkov) [reply source/extensions/filters/http/common/compressor/compressor.cc:502]: Is this function needed? Isn't `end_stream` in `encodeHeaders()` an indication of a header-only response?

Comment (alyssawilk) [reply source/extensions/filters/http/common/compressor/compressor.cc:502]: Yep, that's exactly what I meant with regards to a utility, sorry if that was not clear.   end_stream may also work as a first pass, though I'm not sure if for HTTP/1 it's legal to send a header frame and separate end stream bit for responses without a body.

with regards to guarding, I think rather than having permanent config it'd be fine to just runtime guard - there are instructions in CONTRIBUTING.md  What do you think?

File: api/envoy/extensions/filters/http/compressor/v3/compressor.proto [hunk 3]
```diff
@@ -65,6 +65,10 @@ message Compressor {
     //    To avoid interfering with other compression filters in the same chain use this option in
     //    the filter closest to the upstream.
     bool remove_accept_encoding_header = 3;
+
+    // If true, response would be compressed even if content-length/transfer-encoding header is missing,
+    // unless response has no content (e.g. HEAD, 204 etc)
+    bool always_compress_content_if_exists = 4;
```
[the remark this answers is not in the document]
Comment (tehnerd) [reply api/envoy/extensions/filters/http/compressor/v3/compressor.proto:71]: i can add it there as well. sure

File: source/extensions/filters/http/common/compressor/compressor.cc (hunk 1 above)
Comment (tehnerd) [reply source/extensions/filters/http/common/compressor/compressor.cc:502]: sure. will rework w/ runtime guard.

**Comment by alyssawilk** [comment]
(I'm headed out of office - tagging Yan to take over reviews for the next few work days)

**Comment by tehnerd** [comment]
nvm. will rework. since always_compress_if_content_exists now is a part of common config, there is no need to change signature of isMinimumContentLength

File: source/extensions/filters/http/common/compressor/compressor.cc [hunk 4]
```diff
@@ -506,6 +507,9 @@ bool CompressorFilterConfig::DirectionConfig::isMinimumContentLength(
     }
     return is_minimum_content_length;
   }
+  if (always_compress_content_if_exists_.enabled()) {
+    return true;
+  }
```
Comment (rojkov) [inline source/extensions/filters/http/common/compressor/compressor.cc:512]: This runtime flag is enabled by default which changes the result of this function for existing installations working with HTTP/1.1. Not sure this is a good idea.

If we aim to fix H2 streaming then a better approach would be to narrow the `Transfer-Encoding` check below to apply to HTTP/1.1 only IMHO. `Transfer-Encoding` was introduced in HTTP/1.1 and its `chunked` value was dropped in HTTP2. WDYT?

Comment (tehnerd) [reply source/extensions/filters/http/common/compressor/compressor.cc:512]: i will rework so that default behavior wont be changed and if set to false, it would ignore transfer-encoding chunked and return true unconditionally. how thats sounds @rojkov ?

File: source/extensions/filters/http/common/compressor/compressor.h [hunk 5]
```diff
@@ -188,6 +189,7 @@ class CompressorFilter : public Http::PassThroughFilter {
   bool isAcceptEncodingAllowed(const Http::ResponseHeaderMap& headers) const;
   bool isEtagAllowed(Http::ResponseHeaderMap& headers) const;
   bool isTransferEncodingAllowed(Http::RequestOrResponseHeaderMap& headers) const;
+  bool isResponseWithContent(Http::ResponseHeaderMap& headers) const;
```
Comment (yanavlasov) [inline source/extensions/filters/http/common/compressor/compressor.h:192]: Is this defined anywhere?

File: api/envoy/extensions/filters/http/compressor/v3/compressor.proto [hunk 6]
```diff
@@ -42,6 +42,11 @@ message Compressor {
     // "application/xhtml+xml", "image/svg+xml", "text/css", "text/html", "text/plain", "text/xml"
     // and their synonyms.
     repeated string content_type = 3;
+
+    // Runtime flag that controls whether we compress or not body if no content-length and/or transfer-encoding
+    // headers exist. If set to true (default value) we compress body only if either content-length header exists
+    // (and the value is bigger than the minimum one allowed) or transfer-encoding set to chunked.
+    config.core.v3.RuntimeFeatureFlag do_not_compress_if_no_required_headers = 4;
```
Comment (yanavlasov) [inline api/envoy/extensions/filters/http/compressor/v3/compressor.proto:49]: Does it have to part of the permanent config? I think @alyssawilk suggested moving it to the https://github.com/envoyproxy/envoy/blob/master/source/common/runtime/runtime_features.cc

File: test/extensions/filters/http/common/compressor/compressor_filter_test.cc [hunk 7]
```diff
@@ -233,6 +237,98 @@ TEST_F(CompressorFilterTest, CompressRequest) {
...
+  "compressor_library": {
+     "name": "test",
+     "typed_config": {
+       "@type": "type.googleapis.com/envoy.extensions.compression.gzip.compressor.v3.Gzip"
+     }
+  }
+}
+)EOF");
+  doRequestNoCompression({{":method", "post"}});
+  Http::TestResponseHeaderMapImpl headers{{":method", "post"}};
```
Comment (yanavlasov) [inline test/extensions/filters/http/common/compressor/compressor_filter_test.cc:253]: Nit: It is a bit weird to use the `:method` header in response headers, I guess this is was copied from other tests. If it is not too much trouble can it be changes to `:status` please? Just to reduce confusion a bit.

File: test/extensions/filters/http/common/compressor/compressor_filter_test.cc [hunk 8]
```diff
@@ -233,6 +237,98 @@ TEST_F(CompressorFilterTest, CompressRequest) {
...
+      "@type": "type.googleapis.com/envoy.extensions.compression.gzip.compressor.v3.Gzip"
+    }
+  }
+}
+)EOF");
+  EXPECT_CALL(runtime_.snapshot_, getBoolean("", true)).Times(4);
+  EXPECT_CALL(runtime_.snapshot_, getBoolean("foo_key", true))
+      .Times(1)
+      .WillRepeatedly(Return(false));
+  doRequestCompression({{":method", "post"}}, false);
```
Comment (yanavlasov) [inline test/extensions/filters/http/common/compressor/compressor_filter_test.cc:280]: The test creates 5 byte request body when there is no `content-length` header and the `min_content_length` needed for compression is 30 bytes by default. So the compression should not have been applied, unless I'm missing something.

**Review by yanavlasov — CHANGES_REQUESTED** [review]

File: api/envoy/extensions/filters/http/compressor/v3/compressor.proto (hunk 6 above)
Comment (tehnerd) [reply api/envoy/extensions/filters/http/compressor/v3/compressor.proto:49]: mb i missread her, but my understanding was that it should be protected by RuntimeFlag. as defined here, and by @rojkov comment - it should be disabled by default (so no need to add it into https://github.com/envoyproxy/envoy/blob/master/source/common/runtime/runtime_features.cc#L54). could you folks (@rojkov @alyssawilk) shed a light here please?

File: source/extensions/filters/http/common/compressor/compressor.h (hunk 5 above)
Comment (tehnerd) [reply source/extensions/filters/http/common/compressor/compressor.h:192]: forget to remove. will do

File: test/extensions/filters/http/common/compressor/compressor_filter_test.cc (hunk 7 above)
Comment (tehnerd) [reply test/extensions/filters/http/common/compressor/compressor_filter_test.cc:253]: sure. as i will bake another diff w/ removing `isResponseWithContent` anyway.

File: test/extensions/filters/http/common/compressor/compressor_filter_test.cc (hunk 8 above)
Comment (tehnerd) [reply test/extensions/filters/http/common/compressor/compressor_filter_test.cc:280]: thats the whole point of the diff. new runtime flag would allow to unconditionally compress data if no content-length header and transfer-coding exists. in our deployment we found some h2 servers which do send replies w/o such headers (this is acceptable by h2 rfc). unfortunately there is no visible (for me at least) way to pass that info from h2 frame (where length is specified in h2 world). as a side effect yes, it would compress data even if total lengths is less that min_content_length (but only in scenarios when content-length header is missing)

File: api/envoy/extensions/filters/http/compressor/v3/compressor.proto (hunk 6 above)
Comment (yanavlasov) [reply api/envoy/extensions/filters/http/compressor/v3/compressor.proto:49]: My understanding is that runtime guards are defined here https://github.com/envoyproxy/envoy/blob/master/source/common/runtime/runtime_features.cc I do not know use cases for defining them in the config proto.

File: test/extensions/filters/http/common/compressor/compressor_filter_test.cc (hunk 8 above)
Comment (yanavlasov) [reply test/extensions/filters/http/common/compressor/compressor_filter_test.cc:280]: Ok, I see. Conceptually you could check the minimum length in the encode/decodeData, right? If however `min_content_length` is ignored at least it should be documented in the config proto. Or even the field entirely deprecated.

**Review by yanavlasov — COMMENTED** [review]
/wait

File: test/extensions/filters/http/common/compressor/compressor_filter_test.cc (hunk 8 above)
Comment (tehnerd) [reply test/extensions/filters/http/common/compressor/compressor_filter_test.cc:280]: min_content_length is used for http/1.1 usecases (at least from my understanding @rojkov wanted for it to continue to do so) so there is no point to depreciate it. it is documented that by default we compress only if transfer-encoding exists and set to chunked and/or content-length is set. anyway is it ok w/ @rojkov  i can rework this to be runtime guard. let me know

Comment (rojkov) [reply test/extensions/filters/http/common/compressor/compressor_filter_test.cc:280]: `min_content_length` is a setting related to performance: compressing too small payloads doesn't make sense performance-wise. That's why I think deprecating this field would be too early. Checking it in encode/decodeData is too late, because we need to advertise `content-encoding` before that.

I've just checked the http/1.1 RFC and it says that
```
     If a Content-Length header field (section 14.13) is present, its
     decimal value in OCTETs represents both the entity-length and the
     transfer-length. The Content-Length header field MUST NOT be sent
     if these two lengths are different (i.e., if a Transfer-Encoding
     header field is present). If a message is received with both a
     Transfer-Encoding header field and a Content-Length header field,
     the latter MUST be ignored.
```

There's only one mention that `content-length` is a must in requests with bodies if the server is not known to support http/1.1.

So, basically we should first check `transfer-encoding` containing `chunked` and return true from `isMinimumContentLength()` if that's the case. Then check if `content-length` is present. If it is not then return true. Otherwise return the result of `content-length` >= `min_content_length`. Something like this
```c++
bool CompressorFilterConfig::DirectionConfig::isMinimumContentLength(
    const Http::RequestOrResponseHeaderMap& headers) const {
  if (StringUtil::caseFindToken(headers.getTransferEncodingValue(), ",",
                                Http::Headers::get().TransferEncodingValues.Chunked)) {
    return true;
  }

  const Http::HeaderEntry* content_length = headers.ContentLength();
  if (content_length == nullptr) {
    // FIXME: in case of HTTP/1.0 requests we should probably return false here, because the request is invalid
    //        and should not be touched by the compressor. A request's protocol can be consulted with
    //        decoder_callbacks_->streamInfo().protocol().
    return true;
  }

  uint64_t length;
  const bool is_minimum_content_length =
      absl::SimpleAtoi(content_length->value().getStringView(), &length) &&
      length >= min_content_length_;
  if (!is_minimum_content_length) {
    stats_.content_length_too_small_.inc();
  }

  return is_minimum_content_length;
}
```

Since it changes the current logic it's better to preserve the old code of `isMinimumContentLength()` and continue to use it unless the runtime guard allows the fix.

No need to explore other headers for detecting body-less responses because it will corrupt `Vary` headers.

Comment (yanavlasov) [reply test/extensions/filters/http/common/compressor/compressor_filter_test.cc:280]: Ok makes sense. I have just reread the comment for the `min_content_length` and does say that it is applicable to the content-length value. Thanks for clarification, please disregard my first comment in this thread.

**Comment by tehnerd** [comment]
i've moved new logic to runtime flags. just want to make sure that we are on the same page with everyone

**Comment by rojkov** [comment]
Thanks! Looks good to me.

Not sure if the runtime flag should be enabled by default though to comply with the breaking change policy. In any case this change needs to be listed in `docs/root/version_history/current.rst`.

Also please edit the PR's description.

**Comment by tehnerd** [comment]
updated current.rst . as for default value - the change does not seems super invasive and imho it would be better for it to be enabled by default so more people would be able to test it (and not just us, who found this issue in a first place. not sure that a lot of people enables runtime guarded features by default). going to update PR description in a few

File: source/extensions/filters/http/common/compressor/compressor.cc [hunk 9]
```diff
@@ -506,9 +510,11 @@ bool CompressorFilterConfig::DirectionConfig::isMinimumContentLength(
     }
     return is_minimum_content_length;
   }
-
-  return StringUtil::caseFindToken(headers.getTransferEncodingValue(), ",",
-                                   Http::Headers::get().TransferEncodingValues.Chunked);
+  if (Runtime::runtimeFeatureEnabled(
+          "envoy.reloadable_features.enable_compression_without_chunked_header")) {
+    return true;
```
Comment (rojkov) [inline source/extensions/filters/http/common/compressor/compressor.cc:515]: Please put a comment here that we are going to return true to account for HTTP/2

**Review by rojkov — COMMENTED** [review]
Alright.

To fix the CI coverage check you need to merge the latest HEAD.

**Review by rojkov — APPROVED** [review]
Good. I think this is ready for @envoyproxy/senior-maintainers review.

File: source/extensions/filters/http/common/compressor/compressor.cc (hunk 9 above)
Comment (mattklein123) [reply source/extensions/filters/http/common/compressor/compressor.cc:516]: Is it worth only doing this in the case of HTTP/2? Or should we do it for HTTP/1 also? I don't feel strongly about it but was wondering if the RFC says anything about this one way or the other?

**Review by mattklein123 — COMMENTED** [review]
Thanks LGTM with one small question.

/wait-any

**Comment by tehnerd** [comment]
well current implementation (this diff) will try to do it for both http/1.x and http/2 (as there is no way to distinct them from the filter's POV). Content-Length is going to be used as a hint (e.g. as @rojkov mentioned for performance optimization, when we do not want to compress messages if they are to small), but the part of "either transfer-encoding=chunked  (which was dropped in h/2) or content-length exists in the messages" wont be mandatory anymore.

File: source/extensions/filters/http/common/compressor/compressor.cc (hunk 9 above)
Comment (rojkov) [reply source/extensions/filters/http/common/compressor/compressor.cc:516]: Basically the RFC says that any combination of `transfer-encoding: chunked` and `content-length` is valid. If both are present then the latter must be ignored. The lack of both is valid too with one exception: "a user agent that sends a request containing a message body MUST send a valid Content-Length header field if it does not know the server will handle HTTP/1.1 (or later) requests".  This is a rudiment of HTTP/1.0.

So, the current code in master disables compression for messages that are valid, but look suspicious. And in case of HTTP/2 there is no ground for any suspicions.

Comment (mattklein123) [reply source/extensions/filters/http/common/compressor/compressor.cc:516]: I think we could figure out a way to determine if the request is HTTP/1 or HTTP/2, but if both are valid in both cases this is fine with me. Thanks!

**Review by mattklein123 — APPROVED** [review]
Thanks!

**Comment by rgs1** [comment]
This apparently breaks web sockets for us...

cc: @tehnerd

**Comment by mattklein123** [comment]
@rgs1 feel free to revert while we sort out whatever the issue is.

**Comment by alyssawilk** [comment]
I think the "problem" is that ws doesn't do content length, it does chunk encoding.  So with this change if there's a zip filter configured on a route with ws, it'll zip the ws payload.  Flipping the flag should ideally be sufficient, but I think roll back or fix, we should not compress the payload of upgrades.

**Comment by rgs1** [comment]
@tehnerd @mattklein123 ah ok so here's the problem:

1) our web socket service doesn't set content-type, so we pass isContentTypeAllowed() per https://github.com/envoyproxy/envoy/blob/master/source/extensions/filters/http/common/compressor/compressor.cc#L478

2) after this change, because we don't have the transfer-encoding header, compression kicks in

Chorme is then unhappy and so is Firefox, whereas Safari is still happy.

Who's at fault? Is it my backend service for not setting content-type for a web socket request or should we (the gzip/compressor filter) skip responses that are 101/web-socket/etc?

**Comment by rgs1** [comment]
cc: @bpalermo

**Comment by mattklein123** [comment]
Either way per @alyssawilk I don't think we can break the existing behavior. Can we revert this while we decide what to do unless the fix is obvious and can be done quickly?

**Comment by alyssawilk** [comment]
I can send out a rollback

**Comment by tehnerd** [comment]
i think we probably want to revert this first. As fix is not that obvious (at least for me; how to plug this ws specific condition into compression filter) and would take a while to figure out how to do so w/o layer violation

**Comment by tehnerd** [comment]
just to make sure (as i'm not familiar with WS), @rgs1 the issue is only with compressed "101 Switching Protocols" ? the rest is ok

**Comment by rgs1** [comment]
Here we go #14567 -- thanks!

**Comment by rgs1** [comment]
> just to make sure (as i'm not familiar with WS), @rgs1 the issue is only with compressed "101 Switching Protocols" ? the rest is ok

Yes, I believe that's the only thing that broke for us. Responses like this one:

```
HTTP/1.1 101 Web Socket Protocol Handshake
Upgrade: WebSocket
Connection: Upgrade
Sec-WebSocket-Origin: yada
```

got compressed and Chrome became unhappy ("A server must not mask any frames that it sends to the client").

**Comment by tehnerd** [comment]
interesting. thought that compression should not compress headers itself and seems like 101 is headers only

**Comment by alyssawilk** [comment]
Yeah, technically the payload following the 101 isn't HTTP body, but we have the same API mechanism for upgrade payload data and http body data (onData for both types of data)

I think the upgrade response for HTTP/2 is 200 so the isUpgrade utility might be a better check than status codes.   You can just run an integration test with websocket payload crib off test/integration/websocket_integration_test.cc) and the zip filter, and make sure the payload is not transformed for HTTP/1 or HTTP/2

**Comment by tehnerd** [comment]
thanks for the advice. will look into it.
